subsys/bluetooth/host/conn.c

```diff
@@ -1223,13 +1245,33 @@ void bt_conn_recv(struct bt_conn *conn, struct net_buf *buf, uint8_t flags)
 
 		break;
 	case BT_ACL_CONT:
-		if (!conn->rx_len) {
+		if (!conn->rx) {
 			BT_ERR("Unexpected L2CAP continuation");
 			bt_conn_reset_rx_state(conn);
 			net_buf_unref(buf);
 			return;
 		}
 
+		if (!buf->len) {
+			BT_DBG("Empty ACL_CONT");
+			net_buf_unref(buf);
+			return;
+		}
+
+		/* Have we received the full length field in the L2CAP header
+		 * yet?
+		 */
+		if (conn->rx->len < sizeof(uint16_t)) {
+			uint16_t len = (conn->rx->data[0] | buf->data[0] << 8);
```

Comment:
Not following this one, we will need to keep appending data until there are at least 2 bytes since the start could have 0 bytes there could be necessary to have 2+ rounds to read the entire header if the controller send them one byte at time.

Reply:
This logic is a bit convoluted, I agree. The possible scenarios are:

- 0-byte START => drop buffer on L1214.
- 1-byte START, 0-byte CONT => drop continuation fragment on L1257.
- 1-byte START, >=1-byte CONT => calculate the `conn->rx_len` as done here
- >1-byte START => Normal case

Any suggestions on how to make it clearer? We can at least move the check for empty buf outside the switch.

Reply:
If we drop the buffer at L1214 that means the conn->rx won't be initialized so CONT will cause the Unexpected L2CAP continuation which I guess that needs fixing if we really intend to make this generic. I was actually thinking if it wouldn't be better to reassemble first, problem is there is no flag to indicate end of fragments like there is for ISO which means we need to reconstruct the length while parsing the fragments, perhaps we need to having a separate helper function that does the appending would be better here which we keep calling until there is enough data to set the rx_len.

Reply:
Right, I see. I'll see if I can clean the logic up a bit.

Reply:
I added a suggested re-write in https://github.com/zephyrproject-rtos/zephyr/pull/27169/commits/145257454d3bcd17b206b182878aabebda5e6cc3. It should clean up the logic quite a bit.

Reply:
Did you have a look @Vudentz ? I removed the `[wip]` tag from the commit to let CI run, the commits can be squashed later.
https://github.com/zephyrproject-rtos/zephyr/pull/27169/commits/4adcaafe0b1ccb0bfc3f0968c47c0b7b8e0c9c01